src/common/memory/MemoryTracker.h

```diff
@@ -0,0 +1,143 @@
+ *            occurs, it first tries request size from local reserved.
+ *    Global: Counting the global used memory, the actually used memory may less than the counted
+ *            usage, since threads have quota reservation.
+ */
+class MemoryStats {
+ public:
+  static MemoryStats& instance() {
+    static MemoryStats stats;
+    return stats;
+  }
+
+  /// Inform size of memory allocation
+  inline ALWAYS_INLINE void alloc(int64_t size) {
+    int64_t willBe = threadMemoryStats_.reserved - size;
+
+    if UNLIKELY (willBe < 0) {
+      // if local reserved is not enough, calculate how many bytes needed to get from global.
+      int64_t getFromGlobal = kLocalReservedLimit_;
+      while (willBe + getFromGlobal <= 0) {
+        getFromGlobal += kLocalReservedLimit_;
+      }
+      // allocGlobal() may end with bad_alloc, only invoke allocGlobal() once (ALL_OR_NOTHING
+      // semantic)
+      allocGlobal(getFromGlobal);
+      willBe += getFromGlobal;
+    }
+    // Only update after successful allocations, failed allocations should not be taken into
+    // account.
+    threadMemoryStats_.reserved = willBe;
+  }
+
+  /// Inform size of memory deallocation
+  inline ALWAYS_INLINE void free(int64_t size) {
+    threadMemoryStats_.reserved += size;
+    // Return if local reserved exceed limit
+    while (threadMemoryStats_.reserved > kLocalReservedLimit_) {
+      // freeGlobal() never fail, can be invoked multiple times.
+      freeGlobal(kLocalReservedLimit_);
+      threadMemoryStats_.reserved -= kLocalReservedLimit_;
+    }
+  }
+
+  /// Free global memory, two user case may call this function:
+  /// 1. free()
+  /// 2. destruction of ThreadMemoryStats return reserved memory
+  inline ALWAYS_INLINE void freeGlobal(int64_t bytes) {
+    used_.fetch_sub(bytes, std::memory_order_relaxed);
+  }
+
+  /// Set limit (maximum usable bytes) of memory
+  void setLimit(int64_t limit) {
+    if (this->limit_ != limit) {
+      this->limit_ = limit;
+    }
+  }
+
+  /// Get limit (maximum usable bytes) of memory
+  int64_t getLimit() {
+    return this->limit_;
+  }
+
+  /// Get current used bytes of memory
+  int64_t used() {
+    return used_;
+  }
+
+  /// Calculate used ratio of memory
+  double usedRatio() {
+    return used_ / static_cast<double>(limit_);
+  }
+
+  std::string toString() {
+    return fmt::format("MemoryStats, limit:{}, used:{}", limit_, used_);
+  }
+
+ private:
+  inline ALWAYS_INLINE void allocGlobal(int64_t size) {
+    int64_t willBe = size + used_.fetch_add(size, std::memory_order_relaxed);
+    if (willBe > limit_) {
+      // revert
+      used_.fetch_sub(size, std::memory_order_relaxed);
+      throw std::bad_alloc();
+    }
+  }
+
+ private:
+  // Global
+  int64_t limit_{std::numeric_limits<int64_t>::max()};
+  std::atomic<int64_t> used_{0};
+  // Thread Local
+  static thread_local ThreadMemoryStats threadMemoryStats_;
```

Comment:
For   static thread_local ThreadMemoryStats threadMemoryStats_;
Read some tech spec about thread_local, I think this thread_local should not have cache miss issues;
thread_local is stored in .tdata section in the executable, at runtime, a new data section for each thread is created with a copy of the data in the (read-only) .tdata section, and stored in .tbss section,  so this ThreadMemoryStats should be in each threads .tbss section. see https://akkadia.org/drepper/tls.pdf for details

For limit_ and used_, yes, they do share the same cache line
For limit_, most time it is read only, update logic will first check the new limit is not equls to the old, otherwise, it will not be written;
For used_, yes, it will be read and write by multiple thread, we use the ThreadMemoryStats to lower the frequency to touch this hot address.

Reply:
If many threads read `limit_` when `used_` is modified by some of them, it is still possible to cause false sharing and affect performance. 
Also, I found that `used_` was modified in both `MemoryStats::alloc` and `MemoryStats::free`, which means it could still be a hot address.
@codesigner WDYT?